### PR TITLE
add isInsufficientMaterial and game tags

### PR DESCRIPTION
**Pull request by vladmandic**
add new method `position.isInsufficientMaterial()`  
simple definition is if both sides have either just kings or king+knight or king+bishop on the board  
as those positions cannot result in a checkmate.

add new optinal `game.tag` record which allows setting custom game tags which are then written to pgn when using `pgnWrite`

example usage for both:
```js
    if (position.isStalemate()) {
      game.tag('Termination', 'stalemate');
      game.result('1/2-1/2');
    }
    if (position.isInsufficientMaterial()) {
      game.tag('Termination', 'insufficient material');
      game.result('1/2-1/2');
    }
```

### CODE REVIEW

File: src/private_position/move_generation.ts [hunk 1]
```diff
@@ -143,6 +143,31 @@ export function isStalemate(position: PositionImpl) {
 }
 
 
+/**
+ * Whether the given position is level and there is insufficient material on board.
+ */
+export function isInsufficientMaterial(position: PositionImpl) {
```
Comment (yo35) [inline src/private_position/move_generation.ts:149]: Unit tests should be suitable for this method, in order to ensure it's behavior is correct for standard and corner cases. For example, one corner case: `k7/n1K5/B7/8/8/8/8/8 w - - 0 1` -> I guess your method returns `true`, while it should return `false` (since there is `Bb7#`).

Also: game variants `NO_KING`, `*_KING_ONLY` and `HORDE` should be handled specifically I think (e.g. at Horde, white cannot checkmate with only a queen).

File: src/game.ts [hunk 2]
```diff
@@ -88,6 +90,24 @@ export class Game {
 		}
 	}
 
+	/**
+   * Get or Set custom game tag
+   */
+	tag(tagName: string, value?: string): string | undefined {
```
Comment (yo35) [inline src/game.ts:96]: I don't want a generic key-value tag for the reasons explained in #33.

File: src/private_pgn/pgn_write_impl.ts [hunk 3]
```diff
@@ -229,6 +229,9 @@ export function writeGame(game: Game) {
 	result += writeOptionalHeader('Variant', formatVariant(variant));
 	result += writeOptionalIntegerHeader('WhiteElo', game.playerElo('w'));
 	result += writeOptionalHeader('WhiteTitle', game.playerTitle('w'));
+	for (const [tag, value] of Object.entries(game.tags())) {
```
Comment (yo35) [inline src/private_pgn/pgn_write_impl.ts:232]: Whatever the solution chosen to handle the `Termination` tag, `pgn_read` must be modified as well, for consistency.

**Review by yo35 — CHANGES_REQUESTED** [review]

File: src/game.ts (hunk 2 above)
Comment (vladmandic) [reply src/game.ts:96]: Continuing discussion in #33

File: src/private_position/move_generation.ts (hunk 1 above)
Comment (vladmandic) [reply src/private_position/move_generation.ts]: > For example, one corner case: k7/n1K5/B7/8/8/8/8/8 w - - 0 1 -> I guess your method returns true, while it should return false (since there is Bb7#).

Its a bit trickier since black cannot be forced into such position.  

FIDE declares automatic draw due to insufficient materials when no forced mate is possible.
While USCF allows for any mate, even if its by mistake

For example two knights vs lone king is considered draw under USCF as there can be no forced mate, but not under FIDE as mate is still possible by accident.

I've added a flag `isinsufficientMaterial(forcedMate?: boolean)` to allow both options.  
Notes are in the code.  

> Also: game variants NO_KING, *_KING_ONLY and HORDE should be handled specifically I think (e.g. at Horde, white cannot checkmate with only a queen).

My notes explicitly state that method returns always false for **horde** and **antichess**  
This is similar exception you have in some other methods as well

Comment (yo35) [reply src/private_position/move_generation.ts]: I was not aware of the USCF specificity. Having a flag to support both rules is relevant then.

**Comment by vladmandic** [comment]
closing the pr as new one with separate implementation for just `isInsufficientMatieral` will be created.
